File: exercises/Queue/Queue.cpp
#include <iostream>

class Queue {

private:
  int rear;
  int front;
  int size;
  int *arr;
  int currentSize;

public:
  Queue(int size);
  void Enqueue(int item);
  void Dequeue();
  bool Is_full();
  bool Is_empty();
  void Show_queue();
};

Queue::Queue(int size) {
  this->size = size;
  this->arr = new int[size];
  this->front = 1;
  this->rear = 0;
  this->currentSize = 0;
}

void Queue::Enqueue(int item) {
  if (Is_full())
    std::cout << "Overflow\n";

  else {
    rear = (rear + 1) % size;
    arr[rear] = item;
    std::cout << item << " added to queue\n";
    currentSize++;
  }
}

void Queue::Dequeue() {
  if (Is_empty())
    std::cout << "Underflow\n";
  else {
    arr[front] = 0;
    front = (front + 1) % size;
    currentSize--;
  }
}

bool Queue::Is_full() {
  // if (currentSize == size)
  if ((rear + 2) % size == front)
    return 1;
  return 0;
}

bool Queue::Is_empty() {
  // if (currentSize == 0)
  if (rear + 1 == front)
    return 1;
  return 0;
}

void Queue::Show_queue() {
  std::cout << "[";
  for (int i = 0; i < size; i++) {
    std::cout << "," << arr[i];
  }
  std::cout << "]\n";
}

int main() {
  int size;
  std::cin >> size;
  Queue a(size + 1);
  a.Enqueue(1);
  a.Enqueue(2);
  a.Enqueue(3);
  a.Enqueue(4);
  a.Enqueue(5);
  a.Enqueue(6);
  a.Enqueue(7);
  a.Dequeue();
  a.Dequeue();
  a.Dequeue();
  a.Enqueue(8);
  a.Show_queue();
  return 0;
}

File: exercises/Queue/Queue.py
class Queue:
    def __init__(this, maxsize):
        this.queueArray = [None for i in range(maxsize)]
        #-------------------------------
        # or you could say :
        # for i in range(10):
        #   this.queueArray.append(None)
        #-------------------------------
        this.maxsize = maxsize
        this.rear = 0
        this.front = 1


    def Enqueue(this):
        if(this.Is_full()):
            print("queue is full")
            print(this.queueArray)
        else:
            this.rear += 1
            item = input("item to add :")
            this.queueArray[(this.rear%this.maxsize)] = item
            print("element added")
            print(this.queueArray)

    def Dequeue(this):
        if(this.Is_empty()):
            print("queue is empty")
        else:
            this.queueArray[this.front] = None
            this.front += 1
            print("removed from queue")

    def Is_full(this):
        if ( (this.rear+2)%(this.maxsize) == (this.front)%(this.maxsize) ) :  return True
        else:   return False

    def Is_empty(this):
        if(this.front == this.rear+1):  return True
        else:   return False

while(True):
    print("1. Create a queue.")
    print("2. Enqueue(add) an element.")
    print("3. Dequeue(remove) an element.")
    print("4. Show the queue.")
    print("5. Exit.")
    option = int(input("choose your option :"))

    if(option == 1):
        qSize = int(input("size of queue :"))
        queue = Queue(qSize+1)

    elif(option == 4):
        print(queue.queueArray)

    elif(option == 5):
        break

    else:
        actions = {2: queue.Enqueue, 3: queue.Dequeue}
        action = actions.get(option)
        action()
